source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::hasCacheControlNoTransform(HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::findToken(cache_control->value().c_str(), ",",
+                                 Http::Headers::get().CacheControlValues.NoTransform.c_str());
+  }
+  return false;
+}
+
+// https://tools.ietf.org/html/rfc7231#page-41
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    for (const auto token :
+         StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", false)) {
+      const auto value = StringUtil::trim(StringUtil::cropRight(token, ";", false));
+      const auto q_value = StringUtil::trim(StringUtil::cropLeft(token, ";", false));
+
+      if (value == Http::Headers::get().AcceptEncodingValues.Gzip) {
+        is_gzip = q_value.empty() ? true : (q_value != ZeroQvalueString);
+      }
+      if (value == Http::Headers::get().AcceptEncodingValues.Identity) {
+        is_identity = q_value.empty() ? false : (q_value == ZeroQvalueString);
+      }
+      if (value == Http::Headers::get().AcceptEncodingValues.Wildcard) {
+        is_wildcard = q_value.empty() ? true : (q_value != ZeroQvalueString);
+      }
+    }
+
+    // rfc7231#5.3.4: An "identity" token is used as a synonym for "no encoding" in order to
+    // communicate when no encoding is preferred.
+    if (!is_identity) {
+      return false;
+    }
+
+    // rfc7231#5.3.4: If the representation has no content-coding, then it is acceptable by
+    // default unless specifically excluded by the Accept-Encoding field stating either
+    // "identity;q=0" or "*;q=0" without a more specific entry for "identity".
+    return is_gzip || is_wildcard;
+  }
+
+  return true;
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    std::string value{StringUtil::cropRight(content_type->value().c_str(), ";")};
```

Comment:
value =

Reply:
We need aggregate or direct initialization here. We can also do a static_cast, but I don't think you want that.

Reply:
OK I'm learning stuff I didn't know.  std::string won't have a string_view ctor until C++18 I think.  But when you init like with braces as you have, it must actually somehow figure out that absl::string_view's std::string operator can be called.  So this looks more concise and elegant than writing
  std::string value = std::string(StringUtil::cropRight(content_type->value().c_str(), ";"));
lgtm.

Reply:
I think I learned that from these tips -> https://abseil.io/tips/1
`...If you need to reference the data later or need to modify the data, you can explicitly convert to a C++ string object using string(my_string_view).
`
I think here is the code that gets invoked:
https://github.com/abseil/abseil-cpp/blob/cf1db73d2ad969f0ebbe6f5ed42678517e493da2/absl/strings/string_view.h#L327

I can change it to the direct initialization if you think that it makes the code more readable.